### PR TITLE
Use new buchets style when generating feature profile (bug 1172487)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1172487

I need to run more tests to make sure it doesn't cause any regressions, but it should be almost ready.

### CODE REVIEW

File: package/iframe/js/features.js [hunk 1]
```diff
@@ -5,30 +5,62 @@ var logger = require('./logger');
...
+FeaturesBitField.prototype.toBase64 = function() {
+    return btoa(String.fromCharCode.apply(null, this.values));
+};
+
+function buildFeaturesPromises(features, navigator) {
+    navigator = navigator || window.navigator;
+    var promises = [];
+    // Execute getFeature('hardware.memory') immediately and store the promise
+    // to be able to re-use it without making the call twice.
+    var memoryPromise = navigator.getFeature('hardware.memory');
```
Comment (diox) [inline package/iframe/js/features.js]: drive-by optimisation !

File: package/iframe/js/main.js [hunk 2]
```diff
@@ -65,7 +65,7 @@ function buildQS(profile) {
     }
 
     if (profile) {
-        qs.push('pro=' + profile);
+        qs.push('pro=' + encodeURIComponent(profile));
```
Comment (diox) [inline package/iframe/js/main.js]: One of the things I want to check. We didn't have issues with the profile before, but now it can contain `'='`, `'/'`, etc. Need to make sure this works transparently.

File: package/iframe/tests/features.test.js [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        });
+    });
+
+    it('generateFeatureProfile with empty array works', function(done) {
+        var features = proxyquire('../js/features', {});
+        var navigator = new MockNavigator();
+
+        features.generateFeatureProfile([], navigator).then(function(profile) {
+            //assert.equal(profile, '7fffffffffff.47.8');
+            //console.log(profile);
```
Comment (diox) [inline package/iframe/tests/features.test.js]: Need to finish this test and add a couple more with a non-empty features array.

File: package/iframe/js/features.js [hunk 4]
```diff
@@ -5,30 +5,62 @@ var logger = require('./logger');
...
-        true, // 'getMobileIdAssertion' in window.navigator || 'api.window.Navigator.getMobileIdAssertion',
-        true, // 'manifest.precompile',
-        ['hardware.memory', 512],
-        ['hardware.memory', 1024],
-        true, // NFC
-        'acl.version', // OpenMobile ACL
-        // Don't add any more as long as bug 1172487 is not fixed, it won't
-        // work correctly.
-    ];
+FeaturesBitField.prototype.get = function(i) {
```
Comment (spasovski) [inline package/iframe/js/features.js]: new favourite job interview function

**Comment by diox** [comment]
This is ready for review.

File: package/iframe/js/features.js [hunk 5]
```diff
@@ -5,30 +5,62 @@ var logger = require('./logger');
...
+    // don't need to know the exact value.
+    true, // 'getMobileIdAssertion' in window.navigator || 'api.window.Navigator.getMobileIdAssertion',
+    true, // 'manifest.precompile',
+    ['hardware.memory', 512],
+    ['hardware.memory', 1024],
+    true, // NFC
+    'acl.version', // OpenMobile ACL
+];
+
+function FeaturesBitField(size, values) {
```
Comment (mstriemer) [inline package/iframe/js/features.js]: `values` appears to be unused.

File: package/iframe/js/features.js (hunk 1 above)
Comment (mstriemer) [inline package/iframe/js/features.js]: Is it worth making this generic by saving promises into an object? Seems like probably no but the code is shorter so here it is.

```js
var getFeaturePromises = Object.create(null);
// ...
  promises.push(new Promise(function(resolve, reject) {
    if (!(feature in getFeaturePromises)) {
      getFeaturePromises[feature] = navigator.getFeature(feature);
    }
    getFeaturePromises[feature].then(function(data) {
      resolve(data >= value);
    });
  }));

File: package/iframe/tests/features.test.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    });
+
+});
+
+describe('features bitfield', function() {
+    before(function () {
+        // Node does not have btoa, and we are running the tests outside a
+        // browser. We need to specify binary encoding, utf-8 would give us
+        // wrong results!
+        global.btoa = function(s) { return new Buffer(s, 'binary').toString('base64'); };
```
Comment (mstriemer) [inline package/iframe/tests/features.test.js]: I'd assign this function to a module local to avoid redefining it.

**Comment by mstriemer** [comment]
Why do the unit tests run in node? Did you try setting up a second karma config to run them in a browser?

**Comment by diox** [comment]
They run in node precisely because I didn't try to set up a second karma config to run them in a browser, I was suffering from Yak Shaving Overflow :) I could try it.

**Comment by mstriemer** [comment]
We can move it to karma later. Not a big deal for this one file but once we start adding more tests we'll want to have them run consistently and the node/browser hacks will probably start piling up.

r+wc

File: package/iframe/js/features.js (hunk 1 above)
Comment (diox) [inline package/iframe/js/features.js]: I didn't want to do something too generic for now because we only have one `getFeature()` call to cache right now.

(btw, `getFeature('hardware.memory')` takes anywhere between 60 and 160 ms on my flame, this is why I cached it)

File: package/iframe/js/features.js (hunk 5 above)
Comment (diox) [inline package/iframe/js/features.js]: yeah, removed it.

File: package/iframe/js/features.js (hunk 1 above)
Comment (chuckharmston) [inline package/iframe/js/features.js]: We should file a bug against platform on that, I can't imagine that this couldn't be drastically improved on the OS side.
